#!/bin/bash
:<<!EOF!
 * @Editors: WangJing
 * @Date: 2020-12-30 21:03:03
 * @LastEditors: Hwrn
 * @LastEditTime: 2021-04-08 21:21:32
 * @FilePath: /metaSC/RiboTaxa/qiime_step0.sh
 * @Description:
!EOF!


function show_usage (){
    echo ""
    echo "Usage: $(basename $0) [-o outpath] [-l label] [-t threads] \\"
    echo "                      forward_primer,reverse_primer \\"
    echo ""
    echo 'Example$'" $(basename $0) -o result -l Bacteria \\"
    echo "    TGCCAGCAGCCGCGGTAA,GGACTACCAGGGTATCTAATCCTGTT"
    echo ""
    echo "Required arguments:"
    echo "    primer | primer given in the format of 'forward_primer,reverse_primer'"
    echo "        forward_primer | Forward PCR primer"
    echo "        reverse_primer | Forward PCR primer"
    echo "     * sequence.qza"
    echo "        sequence file generated by 'qiime tool import' earlier."
    echo "        For example:"
    echo "            qiime tools import \\"
    echo "                --input-path silva_132_99_16S.fna \\"
    echo "                --output-path sequences.qza \\"
    echo "                --type 'FeatureData[Sequence]'"
    echo "     * ref-taxonomy.qza"
    echo "        taxonomy file generated by 'qiime tool import' earlier."
    echo "        For example:"
    echo "            qiime tools import \\"
    echo "                --input-path consensus_taxonomy_7_levels.txt \\"
    echo "                --input-format HeaderlessTSVTaxonomyFormat \\"
    echo "                --output-path ref-taxonomy.qza \\"
    echo "                --type 'FeatureData[Taxonomy]'"
    echo ""
    echo "Optional arguments:"
    echo "    -o outpath | Path to store the output files"
    echo "    -l label | label to mark your primer"
    echo "    -t threads | Threads if needed"
    echo "    -s sequence | select sequence file path if you want"
    echo "    -p taxonomy | select ref-taxonomy file path if you want"
    echo ""
    echo "FILE output:"
    echo "   {outpath}/ref_seqs[-label].qza"
    echo "   {outpath}/classifier[-label].qza"
    echo ""

    return 0
}
sequence=sequences.qza
ref=ref-taxonomy.qza
label=''
threads=4


echo "$0 $*" >&2
set -- `getopt o:s:r:l:t:h "$@"`


while [ -n "$1" ]
do
    case "$1" in
    -h)
        show_usage
        exit ;;
    -o)
        outpath=$2
        mkdir -p $outpath
        echo "output files store in"
        echo "    $(cd $(dirname $outpath); pwd)"/"$(basename $outpath)"
        shift ;;
    -s)
        sequence=$2
        shift ;;
    -r)
        ref=$2
        shift ;;
    -l)
        label="-$2"
        shift ;;
    -t)
        threads=$2
        shift ;;
    --) ;;
    *)
        primer_f=${1%%,*}
        primer_r=${1##*,}
        echo "using primer: f: $primer_f"
        echo "              r: $primer_r"
        break ;;
    esac
    shift
done

qiime feature-classifier extract-reads \
    --i-sequences $sequences \
    --p-f-primer $primer_f \
    --p-r-primer $primer_r \
    --o-reads $outpath/ref_seqs${label}.qza \
    --p-n-jobs $threads \
    --verbose

# You can check your version of scikit-learn by typing the following in your python console:
python <<EOF
import sklearn
print('The scikit-learn version is {}.'.format(sklearn.__version__))
EOF

qiime feature-classifier fit-classifier-naive-bayes \
        --i-reference-reads $outpath/ref_seqs${label}.qza \
        --i-reference-taxonomy $ref \
        --o-classifier $outpath/classifier${label}.qza \
        --verbose

echo "Done."
